File: Week 2/3_greatest_common_divisor/gcd.cpp
#include <iostream>


using namespace std;
int gcd_naive(int a, int b) {
  int current_gcd = 1;
  for (int d = 2; d <= a && d <= b; d++) {
    if (a % d == 0 && b % d == 0) {
      if (d > current_gcd) {
        current_gcd = d;
      }
    }
  }
  return current_gcd;
}

long long gcd_fast(long long a, long long b){
    if (b==0){
        return a;
    }else{
        a = a % b;
        gcd_fast(b, a);
    }
}

int main() {
  long long a, b;
  cin >> a >> b;
  cout << gcd_fast(a, b) <<endl;
  return 0;
}

File: Week 2/3_greatest_common_divisor/gcd.py
# Uses python3

def gcd_naive(a, b):
    current_gcd = 1
    for d in range(2, min(a, b) + 1):
        if a % d == 0 and b % d == 0:
            if d > current_gcd:
                current_gcd = d
    return current_gcd

def gcd_fast(a, b):
    
    if b == 0:
        return a
    else:
        a = a % b
        return gcd_fast(b, a)
        

if __name__ == "__main__":
    x = input()
    a, b = [int(i) for i in x.split()]
    print(gcd_fast(a, b))
